packages/serverpod/lib/src/server/server.dart

```diff
@@ -407,9 +407,110 @@ class Server {
   }
 
   Future<Result> _handleUriCall(
-      Uri uri, String body, HttpRequest request) async {
+    Uri uri,
+    String body,
+    HttpRequest request,
+  ) async {
     var path = uri.pathSegments.join('/');
-    return endpoints.handleUriCall(this, path, uri, body, request);
+    var endpointComponents = path.split('.');
+    if (endpointComponents.isEmpty || endpointComponents.length > 2) {
+      return ResultInvalidParams('Endpoint $path is not a valid endpoint name');
+    }
+
+    // Read query parameters
+    var isValidBody = body != 'null' && body.isNotEmpty;
+    Map<String, dynamic> queryParameters = isValidBody ? jsonDecode(body) : {};
+
+    // Add query parameters from uri
+    queryParameters.addAll(uri.queryParameters);
+
+    String endpointName;
+    String methodName;
+
+    if (path.contains('/')) {
+      // Using the new path format (for OpenAPI)
+      var pathComponents = path.split('/');
+      endpointName = pathComponents[0];
+      methodName = pathComponents[1];
+    } else {
+      // Using the standard format with query parameters
+      endpointName = path;
+      var method = queryParameters['method'];
+      if (method is String) {
+        methodName = method;
+      } else {
+        return ResultInvalidParams(
+          'No method name specified in call to $endpointName',
+        );
+      }
+    }
+
+    // Get the the authentication key, if any
+    String? authenticationKey = queryParameters['auth'];
+
+    late MethodCallSession session;
+
+    Endpoint endpoint;
+    MethodConnector method;
+    Map<String, dynamic> paramMap;
+    try {
+      (endpoint, method, paramMap) =
+          await endpoints.getAuthorizedEndpointMethodConnector(
+        createSessionCallback: (connector) {
+          session = MethodCallSession(
+            server: this,
+            uri: uri,
+            body: body,
+            path: path,
+            httpRequest: request,
+            method: methodName,
+            endpoint: endpointName,
+            queryParameters: queryParameters,
+            authenticationKey: authenticationKey,
+            enableLogging: connector.endpoint.logSessions,
+          );
+          return session;
+        },
+        endpointPath: endpointName,
+        methodName: methodName,
+        parameters: queryParameters,
+        serializationManager: serializationManager,
+      );
```

Comment:
@SandPod @Isakdl Not enthralled by the `createSessionCallback`, but at least it avoids any changes to the Session. The obvious drawback is the `late` modifier to the session which can throw if future developers change the behavior. I was considering this instead, which is more code but should be perfectly robust, WDYT?
```dart
MethodCallSession? maybeSession;

    Endpoint endpoint;
    MethodConnector method;
    Map<String, dynamic> paramMap;
    try {
      (endpoint, method, paramMap) =
          await endpoints.getAuthorizedEndpointMethodConnector(
        createSessionCallback: (connector) {
          maybeSession = MethodCallSession(
            server: this,
            uri: uri,
            body: body,
            path: path,
            httpRequest: request,
            method: methodName,
            endpoint: endpointName,
            queryParameters: queryParameters,
            authenticationKey: authenticationKey,
            enableLogging: connector.endpoint.logSessions,
          );
          return maybeSession!;
        },
        endpointPath: endpointName,
        methodName: methodName,
        parameters: queryParameters,
        serializationManager: serializationManager,
      );
    } on MethodNotFoundException catch (e) {
      return ResultInvalidParams(e.message);
    } on InvalidEndpointMethodTypeException catch (e) {
      return ResultInvalidParams(e.message);
    } on EndpointNotFoundException catch (e) {
      return ResultInvalidParams(e.message);
    } on NotAuthorizedException catch (e) {
      return e.authenticationFailedResult;
    } on InvalidParametersException catch (e, stackTrace) {
      var sessionLogId =
          await maybeSession?.close(error: e, stackTrace: stackTrace);
      return ResultInternalServerError(
          e.toString(), stackTrace, sessionLogId ?? 0);
    }

    MethodCallSession? session = maybeSession;
    if (session == null) {
      return ResultInternalServerError(
          'Session not created', StackTrace.current, 0);
    }
```

Reply:
Suggestion: Couldn't you make session nullable? Then we can close it if it has been created?

Another note here is that we will need to call `close` on the session in all cases if an exception occurs.

Reply:
Not sure I follow; the code snippet above makes it nullable? I take it that's preferred then? Regarding closing it, see this comment: https://github.com/serverpod/serverpod/pull/2676#discussion_r1735597748

Reply:
[640023c](https://github.com/serverpod/serverpod/pull/2676/commits/640023c5e4997b5caa1f3ec9a420324d63e493db)